src/drivers/ros3.jl

```diff
@@ -7,11 +7,24 @@ struct ROS3 <: Driver
     fa::API.H5FD_ros3_fapl_t
 end
 
-function ROS3( version::Integer, auth::Bool, region::AbstractString, id::AbstractString, key::AbstractString )
-    return (ROS3 ∘ API.H5FD_ros3_fapl_t)( version, auth, Base.unsafe_convert(Cstring, region), Base.unsafe_convert(Cstring, id), Base.unsafe_convert(Cstring, key))
+function ROS3(
+    version::Integer,
+    auth::Bool,
+    region::AbstractString,
+    id::AbstractString,
+    key::AbstractString
+)
+    return (ROS3 ∘ API.H5FD_ros3_fapl_t)(
+        version,
+        auth,
+        Base.unsafe_convert(Cstring, region),
```

Comment:
Can this be a regular convert or does it have to be an `unsafe_convert`  ?

Reply:
It needs to be an `unsafe_convert`:

```
julia> convert(Cstring, "hello")
ERROR: MethodError: Cannot `convert` an object of type String to an object of type Cstring
Closest candidates are:
  convert(::Type{Cstring}, ::Union{Ptr{Int8}, Ptr{Nothing}, Ptr{UInt8}}) at ~/src/julia-1.7.1/share/julia/base/c.jl:167
  convert(::Type{T}, ::T) where T at ~/src/julia-1.7.1/share/julia/base/essentials.jl:218
Stacktrace:
 [1] top-level scope
   @ REPL[55]:1

julia> Base.unsafe_convert(Cstring, "hello")
Cstring(0x00007fc5c5e06838)
```

Reply:
Probably the best option is to add them as extra fields to the ROS3 struct?

Reply:
Couldn't we just do `unsafe_convert(Cstring, String(region))`?

Reply:
> Couldn't we just do `unsafe_convert(Cstring, String(region))`?

Yes, it's equivalent:
https://github.com/JuliaLang/julia/blob/99225ab1d11925b8e904b0b12f1ee8b0f6946660/base/c.jl#L196-L198
but we have the API, so you should use it.

Either way, you still need to keep a reference to the resulting `String`.

Reply:
@simonbyrne  is this what you mean? 
```julia
function ROS3(
    version::Integer,
    auth::Bool,
    region::AbstractString,
    id::AbstractString,
    key::AbstractString
)

    regionbuf = Base.cconvert(Cstring, region)
    idbuf = Base.cconvert(Cstring, id)
    keybuf = Base.cconvert(Cstring, key)

    GC.@preserve regionbuf idbuf keybuf begin
         regionstr = Base.unsafe_convert(Cstring, regionbuf)
         idstr = Base.unsafe_convert(Cstring, idbuf)
         keystr = Base.unsafe_convert(Cstring, keybuf)
         ROS3(API.H5FD_ros3_fapl_t(version, auth, regionstr, idstr, keystr))
    end

end
```

Reply:
That still is not going to work. We need to preserve `regionbuf`, `idbuf`, and `keybuf`.

Reply:
I think we need this:

```julia
struct ROS3 <: Driver
    version::Int32
    authenticate::hbool_t
    aws_region::String
    secret_id::String
    secret_key::String
end

function Base.unsafe_convert(API.H5FD_ros3_fapl_t, driver::ROS3)
    API.H5FD_ros3_fapl_t(
        driver.version,
        driver.authenticate,
        Base.unsafe_convert(Cstring, driver.aws_region),
        Base.unsafe_convert(Cstring, driver.secret_id),
        Base.unsafe_convert(Cstring, driver.secret_key),
    )
end
```

`ccall` then should then make sure that ROS3 instance and its member Strings are preserved.

Reply:
I think I missed the original point of this all. Please correct me if I'm still wrong in my thinking  
So the reason we need this is because we  use the ROS3 instance in  further calls  (specifically `h5p_set_fapl_ros3` `h5p_get_fapl_ros3`) and those need to convert the outward driver struct `ROS3` to the internal libhdf5  struct `H5FD_ros3_fapl_t` and right now there's a mismatch in that `ROS3` doesn't agree with the layout of  `H5FD_ros3_fapl_t`?

Reply:
We need to convert the Julia struct to the C struct. Part of the issue is that the Julia struct and its Strings could be garbage collected. Meanwhile, the Cstring is an unmanaged pointer to a null terminated character array.

Anyways, it turns out that the C struct definition here is wrong. The correct definition is

https://github.com/HDFGroup/hdf5/blob/100b22e6c23c44a082fd69b8c05a63c7492083f7/src/H5FDros3.h#L82-L88
```c
typedef struct H5FD_ros3_fapl_t {
    int32_t version;
    hbool_t authenticate;
    char    aws_region[H5FD_ROS3_MAX_REGION_LEN + 1];
    char    secret_id[H5FD_ROS3_MAX_SECRET_ID_LEN + 1];
    char    secret_key[H5FD_ROS3_MAX_SECRET_KEY_LEN + 1];
} H5FD_ros3_fapl_t;
```

In Julia, that would be
```julia
struct H5FD_ros3_fapl_t
    version::Int32
    authenticate::Bool
    aws_region::NTuple{H5FD_ROS3_MAX_REGION_LEN + 1,Cchar}
    secret_id ::NTuple{H5FD_ROS3_MAX_SECRET_ID_LEN + 1, Cchar}
    secret_key::NTuple{H5FD_ROS3_MAX_SECRET_KEY_LEN + 1, Cchar};
end
```

I've started to rewrite this. Also, I think we need to check our definition of `hbool_t`. Currently it is `UInt32`, but I think it probably should be `UInt8`.

Reply:
That's probably not the first time there has been a mismatch in definitions. Do you want to push your changes on this branch?

Reply:
Yes, I will later today. I just started it on a different computer than I am on right now. I didn't quite get the tests to pass yet on the branch I was working on.

Reply:
I quickly saw this comment on the hbool issue https://github.com/aldanor/hdf5-rust/issues/28
but perhaps it doesn't affect us since we dont' use msvc